### DISCUSSION TITLE
Problem "zos_operator" command

### DISCUSSION THREAD

**Discussion by ghost** [ansible-collections/ibm_zos_core#592] [Q&A] [answered]
Hi all,

I'm new in this area and i'm having problems  to execute the module 'zos_operator - ibm_zos_core' (I tried zos_command too) to display an information about TCP. Follow the code bellow and the error:

---------------------------------------------------
provision.yml
---
- name: Playbook to call roles in the host
  hosts: "{{ ambiente }}"

  roles:
    - query_cics
    - tso_command
--------------------------------------------------
Role query_cics
--------------------------------------------------
- name: 'Query CICS'
  collections:
    - ibm.ibm_zos_cics
    - sgh.apg_config
  environment: "{{ environment_vars }}"

  cmci_get:
    cmci_host: "{{ cmci_host }}"
    cmci_port: "{{ cmci_port }}"
    cmci_user: "{{ ansible_user }}"
    cmci_password: " {{ ansible_password }}"
    scheme: http
    context: "{{ context }}"
    scope: "{{ name_cics }}"
    type: CICSRegion
  register: records
---
role zOS_operator
---
- name: zOS operator'
  collections:
    - ibm.ibm_zos_core
    - sgh.apg_config
  environment: "{{ environment_vars }}"

  zos_operator:
    cmd: "D TCPIP"
    verbose: true
    wait_time_s: 5
    wait: true
----------------------------------------------------
Display of error:
---

<desb> ESTABLISH SSH CONNECTION FOR USER: xyzxpto
<desb> SSH: EXEC sshpass -d9 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="xyzxpto"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/xyzxpto/.ansible/cp/bf322e5ee0"' desb'/bin/sh -c '"'"'echo ~xyzxpto && sleep 0'"'"''
<desb> (0, b'/home/xyzxpto\n', b'')
<desb> ESTABLISH SSH CONNECTION FOR USER: xyzxpto
<desb> SSH: EXEC sshpass -d9 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="xyzxpto"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb'/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/xyzxpto/.ansible/tmp `"&& mkdir "` echo /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315 `" && echo ansible-tmp-1671733634.5642738-81534-173543433786315="` echo /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315 `" ) && sleep 0'"'"''
<desb> (0, b'ansible-tmp-1671733634.5642738-81534-173543433786315=/home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315\n', b'')
Using module file /home/f3160882/.ansible/collections/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_operator.py
<desb> PUT /home/f3160882/.ansible/tmp/ansible-local-81465opghjjp4/tmpjmptzlll TO /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/AnsiballZ_zos_operator.py
<desb> SSH: EXEC sshpass -d9 sftp -o BatchMode=no -b - -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="xyzxpto"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' '[desb]'
<desb> (0, b'sftp> put /home/f3160882/.ansible/tmp/ansible-local-81465opghjjp4/tmpjmptzlll /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/AnsiballZ_zos_operator.py\n', b'')
<desbr> ESTABLISH SSH CONNECTION FOR USER: xyzxpto
<desb> SSH: EXEC sshpass -d9 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="xyzxpto"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb '/bin/sh -c '"'"'chmod u+x /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/ /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/AnsiballZ_zos_operator.py && sleep 0'"'"''
<desb> (0, b'', b'')
<desb> ESTABLISH SSH CONNECTION FOR USER: xyzxpto
<desb> SSH: EXEC sshpass -d9 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="xyzxpto"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' -tt desb '/bin/sh -c '"'"'_BPXK_AUTOCVT='"'"'"'"'"'"'"'"'“ON”'"'"'"'"'"'"'"'"' ZOAU_HOME='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz””'"'"'"'"'"'"'"'"' PYTHONPATH='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”/lib”'"'"'"'"'"'"'"'"' LIBPATH='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”/lib:“/pp/python/v3r9m0/pyz/bin/python3.9”/lib:/lib:/usr/lib”'"'"'"'"'"'"'"'"' PATH='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”/bin:“/pp/python/v3r9m0/pyz/bin/python3.9”/bin:/bin:/var/bin”'"'"'"'"'"'"'"'"' _CEE_RUNOPTS='"'"'"'"'"'"'"'"'“FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)”'"'"'"'"'"'"'"'"' _TAG_REDIR_ERR='"'"'"'"'"'"'"'"'“txt”'"'"'"'"'"'"'"'"' _TAG_REDIR_IN='"'"'"'"'"'"'"'"'“txt”'"'"'"'"'"'"'"'"' _TAG_REDIR_OUT='"'"'"'"'"'"'"'"'“txt”'"'"'"'"'"'"'"'"' LANG='"'"'"'"'"'"'"'"'“C”'"'"'"'"'"'"'"'"' /pp/python/v3r9m0/pyz/bin/python3.9 /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/AnsiballZ_zos_operator.py && sleep 0'"'"''
<desb> (1, b'**Cannot set locale: The internationalization variable settings are invalid.**\r\n\xe4\xe1\xe1\x93\x96\x90\x98\xf1\x80\xe8\xc7\xc1\x80\xc3?%%?\xcf\xd1>\xc5\x80_\xc1\xcb\xcb/\xc5\xc1\xcb\x80\xf8\xc1\xca\xc8/\xd1>\x80\xc8?\x80\xc8\xc7\xc1\x80\xd1>\xce?\xc4/\xc8\xd1?>\x80\xc4?__/>\xc0\x80\xca\xcd>\x05\xc8\xd1_\xc1\x80?\xf8\xc8\xd1?>\xcb\x06\x8e\xe4\xe1\xe1\x93\x96\x91\x91\xf1\x80\xe8\xc7\xc1\x80\xca\xcd>\x05\xc8\xd1_\xc1\x80?\xf8\xc8\xd1?>\x80S\xd8\xe6\xe3\xf1<\xe1\xe8\xa0\xe5\x80\xcf/\xcb\x80/>\x80\xd1>\xce/%\xd1\xc0\x80\xca\xcd>\x05\xc8\xd1_\xc1\x80?\xf8\xc8\xd1?>\x80?\xca\x80\xd1\xcb\x80>?\xc8\x80\xcb\xcd\xf8\xf8?\xca\xc8\xc1\xc0\x80\xd1>\x80\xc8\xc7\xd1\xcb\x80\xca\xc1%\xc1/\xcb\xc1\x80?\xc3\x80</>\xc5\xcd/\xc5\xc1\x80\xe1>\xce\xd1\xca?>_\xc1>\xc8\x06\x8e\xe4\xe1\xe1\x93\x96\x90\x95\xf1\x80\xe8\xc7\xc1\x80\xcb\xc8\xca\xd1>\xc5\x80\x1bS\xd8\xb8\x1b\x80\xcf/\xcb\x80\xc3?\xcd>\xc0\x80\xcf\xc7\xc1\xca\xc1\x80/\x80\xc0\xc1%\xd1_\xd1\xc8\xc1\xca\x80\xcf/\xcb\x80\xc1\xcc\xf8\xc1\xc4\xc8\xc1\xc0\x80\xc3?%%?\xcf\xd1>\xc5\x80\xc8\xc7\xc1\x80\xcb\xcd\xc2?\xf8\xc8\xd1?>\xcb\x80\xc3?\xca\x80\xc8\xc7\xc1\x80\xca\xcd>\x05\xc8\xd1_\xc1\x80?\xf8\xc8\xd1?>\x80&|\xeb\xf1\xec\x06\x8e\r\n{"changed": false, "failed": true, "msg": "OperatorCmdError(\' 5 \\"D TCPIP\\" VERB WAIT\', 16, [\'\\\\udcc2\\\\udcd7\\\\udce7\\\\udce6\\\\udcf0\\\\udcf0\\\\udcf0\\\\udcf0\\\\udcc9@\\u0167\\\\udc85\\\\udc83@\\\\udc95\\\\udc96\\\\udca3@\\\\udc86\\\\udc96\\\\udca4\\\\udc95\\\\udc84\\\\x15\', \'Ran:  5 \\"D TCPIP\\" VERB WAIT\'])", "exception": "  File \\"/tmp/ansible_zos_operator_payload_bypk_vh9/ansible_zos_operator_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_operator.py\\", line 169, in run_module\\n  File \\"/tmp/ansible_zos_operator_payload_bypk_vh9/ansible_zos_operator_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_operator.py\\", line 364, in run_operator_command\\n", "invocation": {"module_args": {"cmd": "D TCPIP", "verbose": true, "wait_time_s": 5, "wait": true}}}\r\n', b'Shared connection to desb closed.\r\n')
<desb> Failed to connect to the host via ssh: Shared connection to desb closed.
<desb> ESTABLISH SSH CONNECTION FOR USER: xyzxpto
<desb> SSH: EXEC sshpass -d9 ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="xyzxpto"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb '/bin/sh -c '"'"'rm -f -r /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/ > /dev/null 2>&1 && sleep 0'"'"''
<desb> (0, b'', b'')
**The full traceback is:
  File "/tmp/ansible_zos_operator_payload_bypk_vh9/ansible_zos_operator_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_operator.py", line 169, in run_module
  File "/tmp/ansible_zos_operator_payload_bypk_vh9/ansible_zos_operator_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_operator.py", line 364, in run_operator_command
fatal: [DESB]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "cmd": "D TCPIP",
            "verbose": true,
            "wait": true,
            "wait_time_s": 5
        }
    },
    "msg": "OperatorCmdError(' 5 \"D TCPIP\" VERB WAIT', 16, ['\\udcc2\\udcd7\\udce7\\udce6\\udcf0\\udcf0\\udcf0\\udcf0\\udcc9@ŧ\\udc85\\udc83@\\udc95\\udc96\\udca3@\\udc86\\udc96\\udca4\\udc95\\udc84\\x15', 'Ran:  5 \"D TCPIP\" VERB WAIT'])"
}**

---
Notes
---
I put the two roles just to show that one works (query_cics) and another no (zOS operator)

I'm using this collections level:

ibm.ibm_zos_cics    1.0.1  
ibm.ibm_zos_core    1.4.0 (tried with 1.3.0 too)
ibm.ibm_zos_sysauto 1.0.0  
ibm.ibm_zos_zosmf   2.5.0  

python3-3.6.8-18.el7.x86_64

**Comment by ketankelkar** [comment]
From the below portion of your dump, I'm wondering if your ZOAU_HOME might be pointing at the wrong thing (looks like it might be pointing at pyz?)
specifically:
``` ZOAU_HOME='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”```
dump:
```'"'"'BPXK_AUTOCVT='"'"'"'"'"'"'"'"'“ON”'"'"'"'"'"'"'"'"' ZOAU_HOME='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz””'"'"'"'"'"'"'"'"' PYTHONPATH='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”/lib”'"'"'"'"'"'"'"'"' LIBPATH='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”/lib:“/pp/python/v3r9m0/pyz/bin/python3.9”/lib:/lib:/usr/lib”'"'"'"'"'"'"'"'"' PATH='"'"'"'"'"'"'"'"'““/pp/python/v3r9m0/pyz”/bin:“/pp/python/v3r9m0/pyz/bin/python3.9”/bin:/bin:/var/bin”'"'"'"'"'"'"'"'"' CEE_RUNOPTS='"'"'"'"'"'"'"'"'“FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)”'"'"'"'"'"'"'"'"' TAG_REDIR_ERR='"'"'"'"'"'"'"'"'“txt”'"'"'"'"'"'"'"'"' TAG_REDIR_IN='"'"'"'"'"'"'"'"'“txt”'"'"'"'"'"'"'"'"' TAG_REDIR_OUT='"'"'"'"'"'"'"'"'“txt”'"'"'"'"'"'"'"'"' LANG='"'"'"'"'"'"'"'"'“C”'"'"'"'"'"'"'"'"' /pp/python/v3r9m0/pyz/bin/python3.9 /home/xyzxpto/.ansible/tmp/ansible-tmp-1671733634.5642738-81534-173543433786315/AnsiballZ_zos_operator.py && sleep 0'"'"''```

**Comment by ketankelkar** [comment]
Otherwise, it looks like there's an issue with locale
```1, b'Cannot set locale: The internationalization variable settings are invalid.```
I don't have much experience dealing with locale but some of the other folks do. Most of them are currently on vacation, so it might take longer than usual to get a response
 @ddimatos

**Comment by ddimatos** [comment, marked as answer]
@fabiobncbb thank you for reaching out to us,

First thing I notice is typos in your variables, e.g. `BPXK_AUTOCVT` it should be `_BPXK_AUTOCVT`, this typo with a leading underscore seems to be missing from many of your variables:
1. _CEE_RUNOPTS
2.   _TAG_REDIR_ERR
3.   _TAG_REDIR_IN
4.   _TAG_REDIR_OUT
5.   _ENCODE_FILE_NEW
6.   _ENCODE_FILE_EXISTING


Second, 
@ketankelkar has a point, the environment variables look a bit off to me, notice the excessive quotes , eg you have `BPXK_AUTOCVT='"'"'"'"'"'"'"'"'“ON”'"'"'"'"'"'"'"'"' ` where I have `_BPXK_AUTOCVT=ON `

Here is what a log should look like. 
```
 SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="omvsadm"' -o ConnectTimeout=10 -o 'ControlPath="/Users/ddimatos/.ansible/cp/c210400dbb"' ec33012a.vmec.svl.ibm.com '/bin/sh -c '"'"'export _BPXK_AUTOCVT=ON ; export _CEE_RUNOPTS='"'"'"'"'"'"'"'"'FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)'"'"'"'"'"'"'"'"' ; export _TAG_REDIR_ERR=txt ; export _TAG_REDIR_IN=txt ; export _TAG_REDIR_OUT=txt ; export LANG=C ; export ZOAU_HOME=/zoau/v1.2.1 ; export LIBPATH=/zoau/v1.2.1/lib:/python2/usr/lpp/IBM/cyp/v3r9/pyz/lib:/lib:/usr/lib:. ; export PYTHONPATH=/zoau/v1.2.1/lib ; export PATH=/zoau/v1.2.1/bin:/python2/usr/lpp/IBM/cyp/v3r9/pyz/bin:/bin:
```

Here is an example of my `all.yml`, notice the indentation under `environment_vars`, it matters.
```
PYZ: "/python2/usr/lpp/IBM/cyp/v3r9/pyz/"
ZOAU: "/zoau/v1.1.0-ga"
environment_vars:
  _BPXK_AUTOCVT: "ON"
  ZOAU_HOME: "{{ ZOAU }}"
  JAVA_HOME: "{{ JAVA }}"
  PYTHONPATH: "{{ ZOAU }}/lib"
  LIBPATH: "{{ ZOAU }}/lib:{{ PYZ }}/lib:/lib:/usr/lib:."
  PATH: "{{ ZOAU }}/bin:{{ PYZ }}/bin:/bin:/var/bin"
  _CEE_RUNOPTS: "FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)"
  _TAG_REDIR_ERR: "txt"
  _TAG_REDIR_IN: "txt"
  _TAG_REDIR_OUT: "txt"
  LANG: "C"
  _ENCODE_FILE_NEW: "IBM-1047"
  _ENCODE_FILE_EXISTING: "IBM-1047"
```

You can also find this documented [here](https://github.com/IBM/z_ansible_collections_samples/blob/main/docs/share/zos_core/configuration_guide.md) in more detail.

**Comment by ghost** [comment]
Hi all! 
Thanks for help and pointing the way.

After your observations, that points to environmet_vars, i solved the problem! Was an atualization made by the team responsible for ZOAU and Python instalation, that changed the environmet vars. Now i learned how to see this and change in my aplicattion if is necessary.
